File: src/server/RPC.py
from server import Event
from server import WebSocketServer

procedures={}
def listen(callback):
    if procedures.has_key(callback.__name__):
        print 'ERROR: Procedure ' + callback.__name__ + ' already registered!'
        pass
    procedures[callback.__name__] = callback
    pass

def _RPCHandler(event):
    if procedures.has_key(event['procedure']):
        WebSocketServer.send('RPCAnswer', dict(to=event['from'], result=procedures[event['procedure']](event['parameters'])))
    else:
        print '[warn] RPC procedure not found: ' + event['procedure']
        pass
    pass

Event.listen('RPCCall', _RPCHandler)

File: src/editor/development/app/scripts/rpc.js
'use strict';

var $rpc = (function() {
  var pendingBroadcasts = {};

  function getNewId() {
    function genId() {
      return Math.floor(Math.random()*1e9);
    }

    var id = genId();
    while(pendingBroadcasts.hasOwnProperty(id)) id = genId();
    return id;
  }

  function gotRPCAnswer(answer) {
    pendingBroadcasts[answer.to](answer.result);
    delete pendingBroadcasts[answer.to];
  }

  $event.listen('RPCAnswer', gotRPCAnswer);

  function RPCSystem() { }
  RPCSystem.prototype = {
    call: function(procedureName, parameters, callback) {
      var myId = getNewId();
      pendingBroadcasts[myId] = callback;
      $socket.broadcast('RPCCall', {
        procedure: procedureName,
        from: myId,
        parameters: parameters
      });
    }
  };
  return new RPCSystem();
})();
